### PR TITLE
BatchCursor close regression tests

### PR DESCRIPTION
**Pull request by rozza**
Ensure sessions are returned if the batch cursor onNext call results in
an error.

JAVA-3976

### CODE REVIEW

**Comment by jyemin** [comment]
FYI, static checks are failing.

**Review by jyemin — APPROVED** [review]
LGTM

File: driver-reactive-streams/src/test/functional/com/mongodb/reactivestreams/client/BatchCursorPublisherErrorTest.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    @AfterEach
+    public void tearDown() {
+        if (collection != null) {
+            drop(collection.getNamespace());
+        }
+    }
+
+    @TestFactory
+    @DisplayName("test batch cursors close the cursor if onNext throws an error")
+    List<DynamicTest> testBatchCursorThrowsAnError() {
```
Comment (stIncMale) [inline driver-reactive-streams/src/test/functional/com/mongodb/reactivestreams/client/BatchCursorPublisherErrorTest.java:71]: I tried to see if this test actually catches the problem reported in JAVA-3976, and it does not seem like it does: the test passes on both 0b9292e530013c75a53d36d76fb5018a4dd257cf (immediately before [Fix reactive streams batch cursor regression. (#652)](4af9e9f2a541b8a89eadd7f91124b324015866eb)) and on [Fix reactive streams batch cursor regression. (#652)](4af9e9f2a541b8a89eadd7f91124b324015866eb).

However, if we add `System.out.println("AsyncBatchCursor close");` to `AsyncBatchCursor.close` and enable `DEBUG` logging in `logback-test.xml`, we can see that
- with the version immediately before the fix, `AsyncBatchCursor.close` is not called and `killCursor` commands are not issued (nevertheless, the test passes)
- with the fixed version, `AsyncBatchCursor.close` is called and we can see `killCursor` commands in the logs.

I am not sure what would be the right approach to test this. Maybe using the same technique the specification tests do: monitor the commands sent by the client to verify that `killCursors` commands are sent when expected?

**Review by stIncMale — CHANGES_REQUESTED** [review]

File: driver-reactive-streams/src/test/functional/com/mongodb/reactivestreams/client/BatchCursorPublisherErrorTest.java (hunk 1 above)
Comment (rozza) [reply driver-reactive-streams/src/test/functional/com/mongodb/reactivestreams/client/BatchCursorPublisherErrorTest.java:71]: KillCursors won't be sent for all scenarios.

However, this tests onDispose / onCancel handle the error scenario, removing those calls, makes this test fail as the session is not returned.

File: driver-reactive-streams/src/test/functional/com/mongodb/reactivestreams/client/BatchCursorPublisherErrorTest.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+import static com.mongodb.reactivestreams.client.Fixture.serverVersionAtLeast;
+import static java.lang.String.format;
+import static java.util.Arrays.asList;
+import static java.util.Collections.singletonList;
+import static java.util.stream.IntStream.rangeClosed;
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assumptions.assumeTrue;
+import static org.junit.jupiter.api.DynamicTest.dynamicTest;
+
+
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-reactive-streams/src/test/functional/com/mongodb/reactivestreams/client/BatchCursorPublisherErrorTest.java:46]: This level of style grooming I probably don't ever use cycles thinking about, 1 or two spaces seem fine. There are other examples of it used in the codebase. 

This shouldn't be the focus of a code review and we should definitely look to use a tool to standardize the code base, so we don't waste time and money on non-behavioural code styles. Its just not noticeable in the IDE and doesn't change on a code reformat.

[the remark this answers is not in the document]
Comment (stIncMale) [reply driver-reactive-streams/src/test/functional/com/mongodb/reactivestreams/client/BatchCursorPublisherErrorTest.java:46]: Ross, thanks for the explanation. I will ignore formatting in the future reviews.

[the remark this answers is not in the document]
Comment (rozza) [reply driver-reactive-streams/src/test/functional/com/mongodb/reactivestreams/client/BatchCursorPublisherErrorTest.java:46]: If it really stands out do speak up at least until we can add a linter to police such things for us.  

For stylistic feedback there is a general convention just start the conversation with Nit eg:

> Nit: Whitespace

File: driver-reactive-streams/src/test/functional/com/mongodb/reactivestreams/client/BatchCursorPublisherErrorTest.java (hunk 1 above)
Comment (stIncMale) [reply driver-reactive-streams/src/test/functional/com/mongodb/reactivestreams/client/BatchCursorPublisherErrorTest.java:71]: I understand that we may not need to always issue `killCursor`, but it seems reasonable to expect that we need to always close a `Closeable` resource (`AsyncBatchCursor`). While we do this as of [Fix reactive streams batch cursor regression. (#652)](https://github.com/mongodb/mongo-java-driver/pull/4af9e9f2a541b8a89eadd7f91124b324015866eb), this test does not seem to verify this behavior, and the lack of this behavior is a part of what was reported in JAVA-3976.

On the other hand, spending too much time trying to come up with a test for the aforementioned internal behavior may not be a good use of time. It's good to have this new test as is. LGTM.

**Review by stIncMale — APPROVED** [review]

File: driver-reactive-streams/src/test/functional/com/mongodb/reactivestreams/client/BatchCursorPublisherErrorTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (stIncMale) [reply driver-reactive-streams/src/test/functional/com/mongodb/reactivestreams/client/BatchCursorPublisherErrorTest.java]: 👍 I knew about "Nit:", but I forgot to use it, my fault.

**Pull request merged**